src/notification/index.js

```diff
@@ -41,10 +41,12 @@ export type Notification =
   | {| recipient_type: 'private', sender_email: string |};
 
 export const getNarrowFromNotificationData = (
-  data: ?Notification,
+  data: Notification,
   usersById: Map<number, User>,
 ): Narrow | null => {
-  if (!data || !data.recipient_type) {
+  if (!data.recipient_type) {
+    // TODO this condition is impossible if values rightly-typed;
+    //      why is it here? can it really happen?
```

Comment:
Thanks for the digging! That does at least partially answer the "why" question at a causal (vs. "good reason") level. :-) Looks like that code in turn comes ultimately from 427cf00267b3157693470ab0579be578841dafca.

It's not clear that it ever actually made *sense* -- in fact, I think it's clear that when first introduced, it didn't, because the next thing the code did was look specifically for the value to be either 'stream' or 'private'.

(Because `Object` is basically `any`, I don't think even "to satisfy the type-checker" explains it.)

I guess the "can it happen" then, following up on your observations, comes down to "can we actually get a `data` object that's ill-typed". To which the answer is probably "no on Android, yes on iOS" -- on Android it's constructed in the app's own Kotlin code (so a mismatch is an internal error it'd be reasonable to assert on), but on iOS it's more or less straight from the Zulip server.

Note that it's not only `getInitialNotification` -- there are two call sites of `narrowToNotification`, and the other... is fed ultimately (as far as the JS is concerned) by the parameter passed with a `notificationOpened` "device event". But from there I believe the story is the same: on Android from our own Kotlin code, and on iOS more or less unfiltered from the server.